### PR TITLE
Add GSoC 2021 page

### PR DESCRIPTION
**Pull request by XVilka**
For the GSoC 2021 application

### CODE REVIEW

File: content/gsoc/2021.md [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+### Difficulty
+Hard
+
+### Benefits for the student
+Student will understand modern program analysis problems related to type inference,
+understand low level details on how compilers produce native code from OOP concepts,
+and will meet the most common reverse engineering task in its advanced incarnation.
+
+### Assess requirements for midterm/final evaluation
+- 1st term: Loading classes, methods and inherence information and integrating loaded information into the analysis loop, types, `av`, `t`, and`pf` commands.
```
Comment (ret2libc) [inline content/gsoc/2021.md:96]:
```suggestion
- 1st term: Loading classes, methods and inheritance information and integrating loaded information into the analysis loop, types, `av`, `t`, and`pf` commands.
```
??

File: content/gsoc/2021.md [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+- Final term: Additional type propagation and inference based on the loaded information, documentation
+	in the Rizin book and tests.
+
+### Links/Resources
+- [Type Analysis Improvements Project](https://github.com/rizinorg/rizin/projects/3)
+- [VTable detection and parsing](https://github.com/rizinorg/rizin/issues/416)
+- [Constrained types enhacements](https://github.com/rizinorg/rizin/issues/315]
```
Comment (ret2libc) [inline content/gsoc/2021.md:103]:
```suggestion
- [Constrained types enhacements](https://github.com/rizinorg/rizin/issues/315)
```

File: content/gsoc/2021.md [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+3. Please, note, how much time a day/week you are able to spend on this project.
+4. Specify your timezone, since so we can assign you a mentor in the same one, to ease communication.
+5. Submit your proposal early, not in the last minute!
+6. Be sure to choose a “backup” idea (the second task you want to do), so that conflicts (two students for one task) can be resolved.
+
+# Project Ideas
+
+## Rizin
+
+## Type Analysis Improvements
```
Comment (ret2libc) [inline content/gsoc/2021.md:70]: no mentors listed

File: content/gsoc/2021.md [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+
+### Benefits for the project
+This feature would greatly help during exploits development, and people would be able to ditch mona.py for Rizin ;)
+
+### Assess requirements for evaluation
+- 1st term: Creating the language for defining the ROP chain semantics and integrating it with SMT
+	solver
+- Final term: Working ropchain compiler, covered by tests and documented in the Rizin book.
+
+### Mentors
```
Comment (ret2libc) [inline content/gsoc/2021.md:217]: I can be added here.

**Review by wargio — CHANGES_REQUESTED** [review]

**Review by wargio — APPROVED** [review]

**Review by ITAYC0HEN — APPROVED** [review]

File: content/gsoc/2021.md [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+- [ROPC](https://github.com/pakt/ropc)
+- [exrop](https://github.com/d4em0n/exrop)
+- [roper2](https://github.com/oblivia-simplex/roper2)
+- [mona.py](https://www.corelan.be/index.php/2012/12/31/jingle-bofs-jingle-rops-sploiting-all-the-things-with-mona-v2/) from corelan
+- [Hunting for ROP Gadgets in Style](https://media.blackhat.com/us-13/US-13-Quynh-OptiROP-Hunting-for-ROP-Gadgets-in-Style-WP.pdf) (2012)
+- [dropper](https://github.com/rizlik/dropper) a BARF-based rop chain generator
+- [Materials](http://dustri.org/b/files/hacklu2014_r2_exploitation.tar.xz) about the exloitation workshop at Hack.lu 2014
+- [Slides](https://github.com/XVilka/hacklu) for the exploitation part of workshop at Hack.lu 2015
+- [RzEgg related bugs](https://github.com/rizinorg/rizin/labels/RzEgg)
+
```
Comment (thestr4ng3r) [inline content/gsoc/2021.md:239]: Another proposal:
```markdown

## Bindings for languages other than C/C++

Rizin offers a convenient scripting interface through the rz-pipe APIs, which build upon its
command-based interface. While this reduced interface is beneficial and well-suited for many
scripting tasks, building more complex applications generally requires direct access to the
public C api that Rizin offers.
Using this API is directly possible in C and C++, as it is done in Cutter for example,
but for other languages no generic bindings exist so far.
The goal of this task is to use a bindings generator such as SWIG to expose Rizin's C API
to languages such as Python, Java or OCaml.

### Task
 - Integrate SWIG-generated bindings into Rizin's build system
 - Write SWIG interfaces for all mature parts of Rizin's C API
 - Integrate the Python bindings into Cutter's Python support

### Skills
The student should be comfortable with the C and Python languages, as well as have a deep understanding
of common memory management patterns such as ownership and reference counting.

### Difficulty
Advanced

### Benefits for the student
The student will gain an experience of exposing a C-based API with manual memory management to
high-level, object-oriented languages with automatic memory management.

### Assess requirements for midterm/final evaluation
- 1st term: Bindings can be generated as part of the standard Rizin build system and small parts of the core API are already usable.
- Final term: All relevant parts of the API can be used through bindings and also from within Cutter's Python interpreter.

### Mentors
- thestr4ng3r
- xvilka

### Links/Resources
- [SWIG Website](http://www.swig.org/)
- [SWIG 4.0 Documentation](http://www.swig.org/Doc4.0/SWIGDocumentation.html)
- [Small PoC of bindings generated in Rizin's build system](https://github.com/rizinorg/rizin/pull/285)
- [Article about Rizin's build system design](https://rizin.re/posts/why-meson/)
```

File: content/gsoc/2021.md [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+# Cutter
+
+## Plugins and Python High Level API
+
+We currently don't have API almost for plugin authors to use. We need to improve a lot of things about our Plugins support and take it few steps ahead.
```
Comment (thestr4ng3r) [inline content/gsoc/2021.md:244]:
```markdown
We currently don't have API almost for plugin authors to use. We need to improve a lot of things about our Plugins support and take it few steps ahead.
This task is only about improving the Python interface in Cutter, specifically its graphical user interface components. For a task about exposing Rizin's API for disassembly, analysis and other purposes, see the Rizin bindings task above.
 ```

**Comment by ITAYC0HEN** [comment]
@thestr4ng3r PR will be the best. You don't even have to download hugo for this

**Comment by thestr4ng3r** [comment]
https://github.com/rizinorg/website/pull/21

**Pull request merged**